src/libstore/ssh-store.cc

```diff
@@ -0,0 +1,92 @@
+#include "ssh-store.hh"
+#include "remote-fs-accessor.hh"
+#include "archive.hh"
+#include "worker-protocol.hh"
+#include "pool.hh"
+
+namespace nix {
+
+SSHStore::SSHStore(string uri, const Params & params, size_t maxConnections)
+    : Store(params)
+    , RemoteStore(params, maxConnections)
+    , tmpDir(createTempDir("", "nix", true, true, 0700))
+    , socketPath((Path) tmpDir + "/ssh.sock")
+    , sshMaster(startProcess([&]() {
+                auto key = get(params, "ssh-key", "");
+                if (key.empty())
+                    execlp("ssh", "ssh", "-N", "-M", "-S", socketPath.c_str(), uri.c_str(), NULL);
+                else
+                    execlp("ssh", "ssh", "-N", "-M", "-S", socketPath.c_str(), "-i", key.c_str(), uri.c_str(), NULL);
+                throw SysError("starting ssh master");
+      }))
```

Comment:
It would probably be good to start the SSH master lazily. Otherwise, if you have an SSH store in your `binary-caches' setting, the SSH master will be started for every Nix command, even when the SSH store isn't used.